File: scripts/status.mjs
import { readFileSync, readdirSync } from "node:fs";
import { resolve, join } from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";

export const ROOT = fileURLToPath(new URL("..", import.meta.url));
const ZONES = ["engine", "ai", "ui"];
const STATES = new Set(["todo", "in-progress", "review", "done", "blocked", "cut"]);
const read = path => readFileSync(path, "utf8");
const table = text => text.split(/\r?\n/).filter(line => line.startsWith("|")).map(line =>
  line.replace(/^\||\|$/g, "").split("|").map(cell => cell.trim())
).filter(row => /^[SEAUIFR]\d+[a-z]?$/.test(row[0]));

export function minutes(value) {
  if (!/^\d{1,2}:\d{2}$/.test(value)) throw new Error(`Некорректное время: ${value}`);
  const [h, m] = value.split(":").map(Number);
  if (h >= 24 || m >= 60) throw new Error(`Некорректное время: ${value}`);
  return h * 60 + m;
}

export function loadBoard(root = ROOT) {
  const plan = read(join(root, "docs/PLAN.md"));
  const expected = new Map(table(plan).map(row => [row[0], row]));
  const folder = join(root, "docs/status");
  const files = readdirSync(folder).filter(file => file.endsWith(".md") && file !== "META.md");
  if (files.length !== ZONES.length || ZONES.some(zone => !files.includes(`${zone}.md`))) {
    throw new Error("Нужны только три зональных файла: engine.md, ai.md, ui.md");
  }
  const zones = {};
  for (const zone of ZONES) {
    const seen = new Set();
    zones[zone] = table(read(join(folder, `${zone}.md`))).map(cells => {
      if (cells.length !== 7) throw new Error(`${zone}: ожидается 7 колонок`);
      const [id, task, deadline, status, started, finished, comment] = cells;
      const spec = expected.get(id);
      if (seen.has(id) || !spec) throw new Error(`${zone}: повторный/неизвестный тикет ${id}`);
      if (![zone, "все"].includes(spec[1])) throw new Error(`${zone}: чужой тикет ${id}`);
      if (!STATES.has(status)) throw new Error(`${zone}: неизвестный статус ${status}`);
      if (deadline !== spec[3]) throw new Error(`${zone}: дедлайн ${id} расходится с PLAN.md`);
      seen.add(id);
      return { id, task, deadline, due: minutes(deadline), status, started, finished, comment };
    });
    const missing = [...expected.values()].filter(row => [zone, "все"].includes(row[1]) && !seen.has(row[0]));
    if (missing.length) throw new Error(`${zone}: отсутствуют тикеты ${missing.map(row => row[0]).join(", ")}`);
  }
  return { plan, zones };
}

export function nextCut(plan, grouped) {
  const section = plan.split("## Порядок урезания, если отстаём")[1]?.split("\n## ")[0];
  const line = section?.split("\n").find(line => line.startsWith("1."));
  if (!line) throw new Error("В PLAN.md отсутствует порядок урезания");
  for (const raw of line.split(/\s*→\s*/)) {
    const item = raw.replace(/^\d+\.\s*/, "").replace(/\.$/, "");
    const ids = item.match(/\b[EAUSIFR]\d+[a-z]?\b/g) ?? [];
    const rows = ids.flatMap(id => grouped.get(id) ?? []);
    if (rows.length && rows.every(row => row.status === "cut")) continue;
    const marker = item.includes("таймлайн") ? "cut:timeline" : item.includes("Парето") ? "cut:pareto" : null;
    if (marker && rows.length && rows.every(row => row.comment.toLowerCase().includes(marker))) continue;
    return item;
  }
  return "все пункты уже cut; агент, README и Docker не режем";
}

export function renderBoard(root = ROOT, now = new Date()) {
  const { plan, zones } = loadBoard(root);
  const meta = read(join(root, "docs/status/META.md"));
  const start = /^START\s*=\s*([^#\r\n]+)/m.exec(meta)?.[1]?.trim() ?? "";
  let elapsed = 0;
  let warning = false;
  try {
    elapsed = (now.getHours() * 60 + now.getMinutes() - minutes(start) + 1440) % 1440;
  } catch { warning = true; }
  const lines = warning ? ["ПРЕДУПРЕЖДЕНИЕ: START не заполнен или неверен; T+0:00. Сроки пока не подтверждены."] : [];
  lines.push(`Время: T+${Math.floor(elapsed / 60)}:${String(elapsed % 60).padStart(2, "0")}`);
  const grouped = new Map();
  for (const zone of ZONES) {
    lines.push("", `[${zone}]`, "ID | Задача | Статус | Дедлайн");
    const active = [];
    for (const row of zones[zone]) {
      if (!grouped.has(row.id)) grouped.set(row.id, []);
      grouped.get(row.id).push(row);
      const overdue = elapsed > row.due && !["done", "cut"].includes(row.status);
      lines.push(`${row.id} | ${row.task} | ${row.status}${overdue ? " ⚠" : ""} | T+${row.deadline}`);
      if (row.status === "in-progress") active.push(`${row.id} (начато ${row.started || "не указано"})`);
    }
    lines.push(`В работе: ${active.join(", ") || "нет"}`);
  }
  const groups = [...grouped.values()];
  const done = groups.filter(rows => rows.every(row => row.status === "done")).length;
  const worst = Math.max(0, ...groups.flat().filter(row => !["done", "cut"].includes(row.status)).map(row => elapsed - row.due));
  const verdict = worst > 15 ? "ОТСТАЁМ" : worst > 0 ? "РИСК" : "УСПЕВАЕМ";
  const freeze = grouped.get("F0")?.[0]?.due;
  if (freeze === undefined) throw new Error("Отсутствует тикет фриза F0");
  if (worst > 15) lines.push("", "Режем по списку: " + nextCut(plan, grouped));
  lines.push("", `Вердикт: ${verdict}${warning ? " (условно: START не задан)" : ""}`,
    `Готово: ${(100 * done / grouped.size).toFixed(1)}% (${done}/${grouped.size} уникальных тикетов)`,
    `До фриза: ${Math.max(0, freeze - elapsed)} минут${elapsed >= freeze ? ` (фриз наступил ${elapsed - freeze} минут назад)` : ""}`);
  return lines.join("\n");
}

if (process.argv[1] && import.meta.url === pathToFileURL(resolve(process.argv[1])).href) {
  try {
    const args = process.argv.slice(2);
    if (args.length && (args[0] !== "--root" || args.length !== 2)) throw new Error("Использование: npm run status [-- --root PATH]");
    console.log(renderBoard(args.length ? resolve(args[1]) : ROOT));
  } catch (error) {
    console.error(`Ошибка доски статусов: ${error.message}`);
    process.exitCode = 1;
  }
}

File: scripts/status.py
"""Print the hackathon board using only the Python standard library.

Shared tickets count once; they are done only when every zone marks them done.
For partial cuts of U5 use cut:timeline / cut:pareto in its comment.
"""

import argparse
from collections import defaultdict
from datetime import datetime
from pathlib import Path
import re
import sys

ROOT = Path(__file__).resolve().parents[1]
STATES = {"todo", "in-progress", "review", "done", "blocked", "cut"}


def minutes(value):
    if not re.fullmatch(r"\d{1,2}:\d{2}", value):
        raise ValueError(f"Некорректное время: {value}")
    hour, minute = map(int, value.split(":"))
    if minute >= 60 or hour >= 24:
        raise ValueError(f"Некорректное время: {value}")
    return hour * 60 + minute


def table_rows(text):
    for line in text.splitlines():
        if line.startswith("|"):
            cells = [cell.strip() for cell in line.strip().strip("|").split("|")]
            if cells[0] not in {"ID", "Зона", "Критерий", "---"} and not set(cells[0]) <= {"-", ":"}:
                yield cells


def load_board(root):
    plan = (root / "docs/PLAN.md").read_text(encoding="utf-8")
    expected = {row[0]: row for row in table_rows(plan) if len(row) == 6}
    zones = {}
    for path in sorted((root / "docs/status").glob("*.md")):
        if path.name == "META.md":
            continue
        rows = []
        seen = set()
        for cells in table_rows(path.read_text(encoding="utf-8")):
            if len(cells) != 7:
                raise ValueError(f"{path.name}: ожидается 7 колонок")
            ticket, task, deadline, status, started, finished, comment = cells
            if ticket in seen or ticket not in expected:
                raise ValueError(f"{path.name}: повторный или неизвестный тикет {ticket}")
            if status not in STATES:
                raise ValueError(f"{path.name}: неизвестный статус {status}")
            if expected[ticket][1] not in {path.stem, "все"}:
                raise ValueError(f"{path.name}: тикет {ticket} принадлежит другой зоне")
            if deadline != expected[ticket][3]:
                raise ValueError(f"{path.name}: дедлайн {ticket} расходится с PLAN.md")
            seen.add(ticket)
            rows.append(dict(id=ticket, task=task, deadline=deadline, due=minutes(deadline),
                             status=status, started=started, finished=finished, comment=comment))
        required = {key for key, row in expected.items() if row[1] in {path.stem, "все"}}
        if seen != required:
            raise ValueError(f"{path.name}: отсутствуют тикеты {sorted(required - seen)}")
        zones[path.stem] = rows
    if not {"engine", "ai-api", "ui"}.issubset(zones):
        raise ValueError("Нужны все три файла статусов: engine, ai-api, ui")
    return plan, zones


def next_cut(plan, grouped):
    section = plan.split("## Порядок урезания, если отстаём", 1)[1].split("\n## ", 1)[0]
    line = next(line for line in section.splitlines() if line.startswith("1."))
    for item in re.split(r"\s*→\s*", line):
        item = re.sub(r"^\d+\.\s*", "", item).rstrip(".")
        ids = re.findall(r"\b[EAUSIFR]\d+[a-z]?\b", item)
        rows = [row for ticket in ids for row in grouped.get(ticket, [])]
        if rows and all(row["status"] == "cut" for row in rows):
            continue
        marker = "cut:timeline" if "таймлайн" in item else "cut:pareto" if "Парето" in item else None
        if marker and rows and all(marker in row["comment"].lower() for row in rows):
            continue
        return item
    return "все предусмотренные пункты уже cut; агент, README и Docker не режем"


def render_board(root=ROOT, now=None):
    plan, zones = load_board(root)
    meta = (root / "docs/status/META.md").read_text(encoding="utf-8")
    start = re.search(r"^START\s*=\s*([^#\n]+)", meta, re.MULTILINE)
    warning = None
    try:
        start_minutes = minutes(start.group(1).strip() if start else "")
        current = now or datetime.now()
        elapsed = (current.hour * 60 + current.minute - start_minutes) % (24 * 60)
    except ValueError:
        elapsed = 0
        warning = "ПРЕДУПРЕЖДЕНИЕ: START не заполнен или неверен; T+ считается 0:00. Сроки пока не подтверждены."
    lines = [warning] if warning else []
    lines.append(f"Время: T+{elapsed // 60}:{elapsed % 60:02d}")
    grouped = defaultdict(list)
    for zone in ("engine", "ai-api", "ui"):
        lines.extend(["", f"[{zone}]", "ID | Задача | Статус | Дедлайн"])
        active = []
        for row in zones[zone]:
            grouped[row["id"]].append(row)
            overdue = elapsed > row["due"] and row["status"] not in {"done", "cut"}
            mark = " ⚠" if overdue else ""
            lines.append(f"{row['id']} | {row['task']} | {row['status']}{mark} | T+{row['deadline']}")
            if row["status"] == "in-progress":
                active.append(f"{row['id']} (начато {row['started'] or 'не указано'})")
        lines.append("В работе: " + (", ".join(active) or "нет"))
    done = sum(all(row["status"] == "done" for row in rows) for rows in grouped.values())
    late = {ticket: max(elapsed - row["due"] for row in rows if row["status"] not in {"done", "cut"})
            for ticket, rows in grouped.items()
            if any(elapsed > row["due"] and row["status"] not in {"done", "cut"} for row in rows)}
    worst = max(late.values(), default=0)
    verdict = "ОТСТАЁМ" if worst > 15 else "РИСК" if worst else "УСПЕВАЕМ"
    if worst > 15:
        lines.extend(["", "Режем по списку: " + next_cut(plan, grouped)])
    freeze = next(rows[0]["due"] for ticket, rows in grouped.items() if ticket == "F0")
    lines.extend(["", f"Вердикт: {verdict}" + (" (условно: START не задан)" if warning else ""),
                  f"Готово: {100 * done / len(grouped):.1f}% ({done}/{len(grouped)} уникальных тикетов)",
                  f"До фриза: {max(0, freeze - elapsed)} минут" +
                  (f" (фриз наступил {elapsed - freeze} минут назад)" if elapsed >= freeze else "")])
    return "\n".join(lines)


def main():
    if hasattr(sys.stdout, "reconfigure"):
        sys.stdout.reconfigure(encoding="utf-8")
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--root", type=Path, default=ROOT)
    args = parser.parse_args()
    try:
        print(render_board(args.root))
    except (OSError, ValueError, KeyError, IndexError, StopIteration) as error:
        parser.exit(1, f"Ошибка доски статусов: {error}\n")


if __name__ == "__main__":
    main()
